### ISSUE TITLE
The getrandom() feature check is insufficient on some kernels and triggers assertion failure

### ISSUE THREAD

**Issue by nh2** [systemd/systemd#11974]
**systemd version the issue has been seen with**
> 241

**Used distribution**
> Debian buster on a 3.8 kernel

I encountered some very worrying behaviour on a system I was trying to debootstrap Debian buster on (a Chromebook running a 3.8 kernel):

```
# systemd-machine-id-setup 
Assertion '(size_t) r < n' failed at ../src/basic/random-util.c:134, function genuine_random_bytes(). Aborting.
```

Looking at the code, it becomes immediately apparent that this assertion should really never trigger:
https://github.com/systemd/systemd/blob/e62a7fea757f259eb330da5b6d3ab4ede46400a2/src/basic/random-util.c#L127-L134

The only way that can happen is if `getrandom()` returns more bytes than we asked for.

After some painful debugging session I [found that indeed this is the case](https://github.com/dnschneid/crouton/issues/3914#issuecomment-472050196); my debug `printf` shows:

```
getrandom(p, n=16, flags) = (r=355)
```

This problem disappears when running `systemd-machine-id-setup` under `strace`, in which case we observe:

```
getrandom(p, n=16, flags) = (r=-1)
```

Note that the 3.8 kernel should not be able to support `getrandom()` at all, as it was added in Linux 3.17.

This means that while systemd is not at fault, _something_ is returning garbage data here, and systemd's measure to detect absence of the syscall as done in https://github.com/systemd/systemd/blob/e62a7fea757f259eb330da5b6d3ab4ede46400a2/src/basic/random-util.c#L157-L160

is not working as intended, as `getrandom()` does not return `-1` and `errno=ENOSYS` in this case, but `r=355`.

I do not know yet where the fault lies; it might be

* the C wrapper around the syscall not propagating the nonexistence of the syscall correctly
  * whose wrapper is in place, is it glibc's or the [here in the systemd code](https://github.com/systemd/systemd/blob/v241/src/basic/missing_syscall.h#L77-L122)?
* some odd patches in the Chromebook's Linux kernel that make this non-existent syscall return garbage

But the fact that `strace` "fixes" the issue and makes the syscall appear nonexistent as expected feels extremely suspicious to me, so I thought I'd raise the issue and ask for some help figuring this out.

**Comment by nh2** [comment]
>     * whose wrapper is in place, is it glibc's or the [here in the systemd code](https://github.com/systemd/systemd/blob/v241/src/basic/missing_syscall.h#L77-L122)?

I have checked now that systemd's wrapper https://github.com/systemd/systemd/blob/v241/src/basic/missing_syscall.h#L112

is not involved at all, so it must be the glibc wrapper.

**Comment by poettering** [comment]
so you have a really old kernel but a new glibc that already knows getrandom()?

**Comment by nh2** [comment]
> so you have a really old kernel but a new glibc that already knows getrandom()?

@poettering That's right.

As per https://www.gnu.org/software/libc/manual/html_node/System-Calls.html, I would expect

> If you specify an invalid sysno, `syscall` returns `-1` with `errno = ENOSYS`.

to happen, which `systemd-machine-id-setup` would see and fall back to `urandom`. Funnily, that happens only when running under `strace`.

**Comment by nh2** [comment]
Also note that `r = 355`, is the syscall number of `getrandom()`  on my platform.

So I suspect what happens is that as per kernel calling convention, `355` is placed in register `eax` to indicate the syscall to execute, and it stays there (for successful syscalls the result is left in `eax`), as opposed to being translated to `-1` for failure.

**Comment by poettering** [comment]
so, this appears to be a glibc bug then, no? have you reported that there? can we close this here?

The man page is pretty explicit:

> ENOSYS The glibc wrapper function for getrandom() determined that the underlying kernel does not implement this system call.

**Comment by nh2** [comment]
> can we close this here?

Yes, given that systemd's wrapper isn't active, I don't think there's anything it can do there.

---

I am still puzzled what is going on; if I write the assembly to do that syscall directly, the returned `eax` is `-38` (`-ENOSYS`) as expected.

But both musl and glibc's `syscall()` wrapper seem to lose it somehow, and they both are fixed by `strace`.

The assembly they generate looks sane. I must be missing something.

**Comment by poettering** [comment]
maybe the kernel you are using assigned a different syscall onto getrandom()s system call nr?

**Comment by nh2** [comment]
> maybe the kernel you are using assigned a different syscall onto getrandom()s system call nr?

I've tried also quite-certainly-invalid syscall numbers like `500`, and get the same results on this system (the syscall number itself is returned by `syscall()`, while on other systems that too returns `ENOSYS`). But I haven't located exactly yet where this odd behaviour sits. If I write the register-based syscall assembly directly (using `int 0x80` to trigger it), I can see `-ENOSYS` being returned from the kernel correctly in all those cases. But any of the wrapping, e.g. `syscall()` from both glibc and musl (including when using upstream musl git), apparently discards the `-ENOSYS` somehow (or never makes the call in the first place).

This suggests that the kernel behaves correctly at least for the 0x80 case, and that the libcs are at fault somehow.

The only explanation I can come up with so far is that they perhaps do not use `int 0x80` as my direct-to-the-kernel test does; perhaps they use `sysenter` (via `vsyscall`?) and in that case the kernel actually returns garbage, or it's a rare+bugged code path in the libcs?

I am not sure, need to investigate more.

**Comment by nh2** [comment]
I had some more time just now to look into this, and found something interesting:

As mentioned above, when I use an invalid syscall number with `int 0x80`, the kernel behaves normally.

**But I found that if I make the syscall via `sysenter`**, the second possible way to make a syscall on 32-bit x86 Linux (which is likely the version preferred by Linux, set in the VDSO page, and thus dynamically chosen by glibc), **then I can see this Chromebook's 3.8 kernel misbehave**.

Repro code adapted from [this nice answer](https://reverseengineering.stackexchange.com/questions/2869/how-to-use-sysenter-under-linux/2894#2894) that shows `int 0x80`, `sysenter`, and explains the dynamic VDSO switch:

```c
# Compile and run with:
# gcc -m32 -o sysenter-printf sysenter-printf.S && ./sysenter-printf

.data
msg:
    .ascii "Hello World!\n"
    len = . - msg

msg2:
    .ascii "returned %ld!\n"
    len2 = . - msg2

.text
.globl main
main:
    pushl  %ebp
    movl   %esp, %ebp
# Write the string to stdout
    movl   $len, %edx
    movl   $msg, %ecx
    movl   $1, %ebx
# 4 for write()
    #movl   $4, %eax
# 500 for an invalid syscall
    movl   $500, %eax
# Setting the stack for the systenter
    pushl  $sysenter_ret
    pushl  %ecx
    pushl  %edx
    pushl  %ebp
    movl   %esp,%ebp
    sysenter
# and exit
sysenter_ret:

    pushl %eax
    pushl $msg2
    call printf
    popl %eax
    popl %eax

    movl   $0, %ebx
    movl   $1, %eax
# Setting the stack for the systenter
    pushl  $sysenter_ret # Who cares, this is an exit !
    pushl  %ecx
    pushl  %edx
    pushl  %ebp
    movl   %esp,%ebp
    sysenter
```

The above prints `returned 500!` on the Chromebook 3.8 i686 kernel, and `returned -38!` (`ENOSYS`) on my 4.15 x86_64 Ubuntu 16.04 kernel.

`sysenter` also correctly returns `-38` on an Ubuntu 18.04 i686 4.15 kernel, meaning that this problem isn't happening simply across all i686 kernels, but somewhat specific to either 3.8 or Chromebooks kernels in general.

**So in this case, the 3.8 Chromebook kernel apparently really gives the libcs the wrong information, suggesting that the syscall succeeded when it did not.**

Now the mechanism by which the failure occurs is clear, and I just have to find out if this is a Chromebook specific patch that breaks the kernel, or something else.

**Comment by nh2** [comment]
(Summary at the bottom.)

## Mystery lifted

Browsing through the git history of `arch/x86/kernel` at https://github.com/RaumZeit/linux-chromebook/commits/chromeos-3.8/arch/x86/kernel/, I believe I found the relevant commits.

* https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?h=x86/urgent&id=554086d85e71f30abe46fc014fea31929a7c6a8a
  * date 2014-06-23
* https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?h=x86/urgent&id=8142b215501f8b291a108a202b3a053a265b03dd
  * date 2014-07-22

The first commit introduced the bug, the second one fixed it with commit message:

```
Commit 554086d ("x86_32, entry: Do syscall exit work on badsys
(CVE-2014-4508)") introduced a regression in the x86_32 syscall entry
code, resulting in syscall() not returning proper errors for undefined
syscalls on CPUs supporting the sysenter feature.

The following code:

> int result = syscall(666);
> printf("result=%d errno=%d error=%s\n", result, errno, strerror(errno));

results in:

> result=666 errno=0 error=Success

Obviously, the syscall return value is the called syscall number, but it
should have been an ENOSYS error. When run under ptrace it behaves
correctly, which makes it hard to debug in the wild:

> result=-1 errno=38 error=Function not implemented
```

That is exactly what I described above.

---

### ChromeOS kernels

The above is all upstream kernel problems. Regarding the Chromebook's kernel, today both patches are in ChromiumOS's kernel sub tree:

* https://chromium.googlesource.com/chromiumos/third_party/kernel/+/554086d85e71f30abe46fc014fea31929a7c6a8a
* https://chromium.googlesource.com/chromiumos/third_party/kernel/+/8142b215501f8b291a108a202b3a053a265b03dd

But https://github.com/RaumZeit/linux-chromebook/commits/chromeos-3.8/arch/x86/kernel/ has, cherry-picked on 2015-12-01 (more than 1 year after bug and fix had been written) only the bug-introducing one, not the fix.

Below is a detailed research on which ChromeOS versions have the bug:

* The problem **did not exist** in ChromeOS kernel 3.4:
  * [`chromeos-3.4`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.4/arch/x86/kernel/entry_32.S) (tip `29a2e2836ff9ea65a603c89df217f4198973a74f`)
* The problem **exists** in ChromeOS kernels 3.8, 3.10 and 3.14:
  * [`chromeos-3.8`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.8/arch/x86/kernel/entry_32.S) (tip `f6f45b7aaa932acbb16d2af05daff8eb52281168`)
  * [`chromeos-3.10`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.10/arch/x86/kernel/entry_32.S) (tip `742760d608f9fdc598d67ac3ed1f9407c132adc0`)
  * [`chromeos-3.14`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.14/arch/x86/kernel/entry_32.S) (tip `035c527d3dfb9deac21185238c1e0ffdf1647023`)
* The problem **was fixed** in ChromeOS kernel 3.18:
  * [`chromeos-3.18`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.18/arch/x86/kernel/entry_32.S) (tip `4fcc8687ab594821dd3ad449a24f1b2244d2d675`)

In terms of cherry-picked-bug-time, most interesting is the case of [`chromeos-3.14`](https://chromium.googlesource.com/chromiumos/third_party/kernel/+log/refs/heads/chromeos-3.14/arch/x86/kernel/entry_32.S),

* the last commit was made 1 year ago from now (so I take it was maintained until then)
* the bug was cherry-picked _3 years, 4 months ago_

That means that for > 2 years, a maintained ChromeOS kernel shipped this bug breaking existence-of-syscall detection, and thus `systemd`, and thus [`crouton`](https://github.com/dnschneid/crouton) for newer Linux distros, even though an upstream fix had been available for that duration.

## Summary

* An [upstream Linux bug](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?h=x86/urgent&id=554086d85e71f30abe46fc014fea31929a7c6a8a) broke `ENOSYS` being returned on invalid syscalls if `sysenter` was used.
* It was [fixed in upstream Linux](https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?h=x86/urgent&id=8142b215501f8b291a108a202b3a053a265b03dd) 1 month later (and the bug [apparently](https://patchwork.ozlabs.org/patch/372692/) never made it into an upstream kernel release without the fix).
* ChromiumOS cherry-picked the bug starting from their 3.8 release, but apparently forgot to cherry-pick the fix for years.
* As a result `systemd` based Linux distributions in [`crouton`](https://github.com/dnschneid/crouton) chroots work for any 32-bit `i686` ChromeOS that has `uname -a` < 3.8, break for everything above that until 3.18, where they should work again.
  * Unfortunately, many ChromeOS laptops (including the one I found this on) were end-of-lifed on versions within the broken range. For those, a workaround should be downgrading to a ChromeOS version that has a kernel <= 3.4. (Note that running unmaintained old kernels is generally insecure, but unfortunately booting custom kernels isn't easy on that range of devices either.)
